res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,730 @@
+    });
+};
+
+PioneerDDJ400.startLoopLightsBlink = function(channel, control, status, group) {
+    var blink = 0x7F;
+
+    PioneerDDJ400.stopLoopLightsBlink(group, control, status);
+
+    PioneerDDJ400.timers[group][control] = engine.beginTimer(500, function() {
+        blink = 0x7F - blink;
+
+        // When adjusting the loop out position, turn the loop in light off
+        if (PioneerDDJ400.loopAdjustOut[channel]) {
+            midi.sendShortMsg(status, 0x10, 0x00);
+            midi.sendShortMsg(status, 0x4C, 0x00);
+        } else {
+            midi.sendShortMsg(status, 0x10, blink);
+            midi.sendShortMsg(status, 0x4C, blink);
+        }
+
+        // When adjusting the loop in position, turn the loop out light off
+        if (PioneerDDJ400.loopAdjustIn[channel]) {
+            midi.sendShortMsg(status, 0x11, 0x00);
+            midi.sendShortMsg(status, 0x4E, 0x00);
+        } else {
+            midi.sendShortMsg(status, 0x11, blink);
+            midi.sendShortMsg(status, 0x4E, blink);
+        }
+    });
+
+};
+
+PioneerDDJ400.stopLoopLightsBlink = function(group, control, status) {
+    PioneerDDJ400.timers[group] = PioneerDDJ400.timers[group] || {};
+
+    if (PioneerDDJ400.timers[group][control] !== undefined) {
+        engine.stopTimer(PioneerDDJ400.timers[group][control]);
+    }
+    PioneerDDJ400.timers[group][control] = undefined;
+    PioneerDDJ400.setLoopButtonLights(status, 0x7F);
+};
+
+PioneerDDJ400.loopToggle = function(value, group, control) {
+    var status = group === "[Channel1]" ? 0x90 : 0x91,
+        channel = group === "[Channel1]" ? 0 : 1;
+
+    PioneerDDJ400.setReloopLight(status, value ? 0x7F : 0x00);
+
+    if (value) {
+        PioneerDDJ400.startLoopLightsBlink(channel, control, status, group);
+    } else {
+        PioneerDDJ400.stopLoopLightsBlink(group, control, status);
+        PioneerDDJ400.loopAdjustIn[channel] = false;
+        PioneerDDJ400.loopAdjustOut[channel] = false;
+    }
+};
+
+//
+// CUE/LOOP CALL
+//
+
+PioneerDDJ400.cueLoopCallLeft = function(_channel, _control, value, _status, group) {
+    if (value) {
+        engine.setValue(group, "loop_scale", 0.5);
+    }
+};
+
+PioneerDDJ400.cueLoopCallRight = function(_channel, _control, value, _status, group) {
+    if (value) {
+        engine.setValue(group, "loop_scale", 2.0);
+    }
+};
+
+//
+// BEAT SYNC
+//
+// Note that the controller sends different signals for a short press and a long
+// press of the same button (which means we have to reimplement any short press
+// logic on the long press function).
+//
+
+PioneerDDJ400.syncPressed = function(channel, control, value, status, group) {
+    engine.setValue(group, "sync_enabled", value);
+};
+
+PioneerDDJ400.syncLongPressed = function(channel, control, value, status, group) {
+    engine.setValue(group, "sync_enabled", value); // reimplements short press
+    engine.setValue(group, "sync_enabled", 0x01);  // ensures sync lock
+};
```

Comment:
This doesn't seem correct... I think `syncPressed` should toggle `beatsync` and `syncLongPressed` should set `sync_enabled` to 1 only on button down. If `sync_enabled` is 1 in `syncPressed`, then set `sync_enabled` to 0 without changing `beatsync`.

Reply:
I've found these controls quite confusing (their documentation is a bit vague), but I think I understand how they work now. Changes made.